mlflow/pyfunc/__init__.py

```diff
@@ -828,13 +828,160 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
+                f"return value, but model output '{spec.name}' has shape {spec.shape}.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+    else:
+        raise MlflowException(
+            f"Unknown schema output spec {spec}.", error_code=INVALID_PARAMETER_VALUE
+        )
+
+
+def _infer_spark_udf_return_type(model_output_schema):
+    from pyspark.sql.types import StructType, StructField, ArrayType
+
+    if len(model_output_schema.inputs) == 1:
+        return _convert_model_output_spec_to_spark_type(model_output_schema.inputs[0])
+
+    return StructType(
+        [
+            StructField(
+                name=spec.name or str(i), dataType=_convert_model_output_spec_to_spark_type(spec)
+            )
+            for i, spec in enumerate(model_output_schema.inputs)
+        ]
+    )
+
+
 def _parse_spark_datatype(datatype: str):
     from pyspark.sql.functions import udf
 
     return udf(lambda x: x, returnType=datatype).returnType
 
 
-def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LOCAL):
+def _convert_array_values(values, spark_type, spark_primitive_type_to_np_type):
+    """
+    Convert list or numpy array values to spark dataframe column values.
+    """
+    from pyspark.sql.types import ArrayType
+
+    # Get element type from a array<E> type or a array<array<E>> type.
+    # and get the array_dim value (1 or 2).
+    array_dim = 1
+    elem_type = spark_type.elementType
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+        array_dim = 2
+
+    if type(elem_type) not in spark_primitive_type_to_np_type:
+        raise MlflowException(
+            "Unsupported array type field with element type "
+            f"{elem_type.simpleString()} in struct type.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    np_type = spark_primitive_type_to_np_type[type(elem_type)]
+
+    if array_dim == 1:
+        return [np.array(v, dtype=np_type) for v in values]
+    else:
+        return [list(np.array(v, dtype=np_type)) for v in values]
+
+
+def _get_spark_primitive_types():
+    from pyspark.sql import types
+
+    return (
+        types.IntegerType,
+        types.LongType,
+        types.FloatType,
+        types.DoubleType,
+        types.StringType,
+        types.BooleanType,
+    )
+
+
+def _check_udf_return_struct_type(struct_type):
+    from pyspark.sql.types import StructType, ArrayType
+
+    primitive_types = _get_spark_primitive_types()
+
+    for field in struct_type.fields:
+        field_type = field.dataType
+
+        if isinstance(field_type, primitive_types):
+            pass
+        elif isinstance(field_type, ArrayType):
+            if not _check_udf_return_array_type(field_type, allow_nested_struct=False):
+                return False
+        else:
+            return False
+
+    return True
+
+
+def _check_udf_return_array_type(array_type, allow_nested_struct):
```

Comment:
```suggestion
def _check_udf_return_array_type(array_type, allow_struct):
```

Can we rename `allow_nested_struct` to `allow_struct`? `allow_nested_struct` sounds like `struct<struct<...>>`.